File: leetcode/21-merge-two-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        ListNode *head(nullptr), *current(nullptr);
        while (l1 != nullptr && l2 != nullptr) {
            auto to_insert = l1->val < l2->val ? l1 : l2;
            if (l1->val < l2->val) {
                l1 = l1->next;
            } else {
                l2 = l2->next;
            }
            if (head == nullptr) {
                head = to_insert;
            } else {
                current->next = to_insert;
            }
            current = to_insert;
        }
        while (l1 != nullptr || l2 != nullptr) {
            auto to_insert = l1 == nullptr ? l2 : l1;
            if (l1 == nullptr) {
                l2 = l2->next;
            } else {
                l1 = l1->next;
            }
            if (head == nullptr) {
                head = to_insert;
            } else {
                current->next = to_insert;
            }
            current = to_insert;
        }

        return head;
    }
};

File: leetcode/21-merge-two-sorted-lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def merge_two_lists(l1, l2)
  head = ListNode.new 0
  current = head
  while not l1.nil? and not l2.nil?
    if l1.val <= l2.val 
      current.next = l1
      current = l1
      l1 = l1.next
    else
      current.next = l2
      current = l2
      l2 = l2.next
    end
  end
  while not l1.nil?
    current.next = l1
    current = l1
    l1 = l1.next
  end
  while not l2.nil?
    current.next = l2
    current = l2
    l2 = l2.next
  end
  head.next
end
